### PR TITLE
Xav/concurrent cluster resources

### PR DESCRIPTION
**Pull request by xavpaice**
## Description, Motivation and Context

Adds concurrency to the clusterResources collector, in an attempt to speed it up

## Checklist

- [x] New and existing tests pass locally with introduced changes.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] The commit message(s) are informative and highlight any breaking changes
- [ ] Any documentation required has been added/updated. For changes to https://troubleshoot.sh/ create a PR [here](https://github.com/replicatedhq/troubleshoot.sh/pulls)

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by xavpaice** [comment]
Draft PR.   Need to test in multiple environments to determine if it really does make a difference or not.

File: pkg/collect/cluster_resources.go [hunk 1]
```diff
@@ -156,190 +166,144 @@ func (c *CollectClusterResources) Collect(progressChan chan<- interface{}) (Coll
...
-	output.SaveResult(c.BundlePath, "cluster-resources/clusterRoleBindings-errors.json", marshalErrors(clusterRoleBindingsErrors))
+	clusterResourceCollectors = append(clusterResourceCollectors, func() {
+		clusterRoleBindings, clusterRoleBindingsErrors := clusterRoleBindings(ctx, client)
+		output.SaveResult(c.BundlePath, "cluster-resources/clusterRoleBindings.json", bytes.NewBuffer(clusterRoleBindings))
+		output.SaveResult(c.BundlePath, "cluster-resources/clusterRoleBindings-errors.json", marshalErrors(clusterRoleBindingsErrors))
+	})
 
+	// run the collectors
+
+	maxNbConcurrentGoroutines := 5
```
Comment (xavpaice) [inline pkg/collect/cluster_resources.go]: tune me, and put me somewhere more appropriate

File: pkg/collect/cluster_resources.go [hunk 2]
```diff
@@ -156,190 +166,144 @@ func (c *CollectClusterResources) Collect(progressChan chan<- interface{}) (Coll
...
+	concurrentGoroutines := make(chan struct{}, maxNbConcurrentGoroutines)
+
+	for _, thisClusterResourceCollector := range clusterResourceCollectors {
+		wg.Add(1)
+		go func(crCollector getClusterResource) {
+			defer wg.Done()
+			concurrentGoroutines <- struct{}{}
```
Comment (xavpaice) [inline pkg/collect/cluster_resources.go:299]: this pattern kicks off all the goroutines at once, but blocks the actual work till the channel has a free slot

**Comment by banjoh** [comment]
Are we confident that this is the root cause of the slowness? Have we explored request throttling?

I modified client-side rate limits and ran a before and after test on a local cluster with these results

Before
```sh
[evans] $ time ./bin/preflight ./examples/preflight/sample-preflight.yaml --debug

...truncated deprecation warnings (which need fixing)

real    0m6.646s
user    0m0.206s
sys     0m0.112s
```

After
```sh
[evans] $ time ./bin/preflight ./examples/preflight/sample-preflight.yaml --debug
[evans] $ time ./bin/preflight ./examples/preflight/sample-preflight.yaml --debug
  Running Preflight Checks ⠸ I1213 12:14:14.148325   72455 request.go:682] Waited for 1.491075542s due to client-side throttling, not priority and fairness, request: GET:https://0.0.0.0:52414
/apis/apps/v1
  Running Preflight Checks ⠧ I1213 12:14:24.646801   72455 request.go:682] Waited for 11.989650708s due to client-side throttling, not priority and fairness, request: GET:https://0.0.0.0:5241
4/apis/k3s.cattle.io/v1
  Running Preflight Checks ⠋ I1213 12:14:44.147273   72455 request.go:682] Waited for 1.496777292s due to client-side throttling, not priority and fairness, request: GET:https://0.0.0.0:52414
/apis/node.k8s.io/v1beta1
  Running Preflight Checks ⠼ I1213 12:14:54.647183   72455 request.go:682] Waited for 11.996566167s due to client-side throttling, not priority and fairness, request: GET:https://0.0.0.0:5241
4/apis/authentication.k8s.io/v1

...truncated warnings related to client-side throttling

real    2m46.546s
user    0m0.803s
sys     0m0.487s
```

Note the logged warnings on throttled requests. I halved the default rate limiting values and ran this on a cluster that virtually has no load. An API server on production environment might throttle requests as well. Cluster resources collector does collect a lot of information.

Here is the code I changed in [this code location](https://github.com/replicatedhq/troubleshoot/blob/c8820b3d8e5e9b3ec85d4799fe3e90c28ec3e905/pkg/preflight/run.go#L251)

```go
	restConfig, err := k8sutil.GetRESTConfig()
	if err != nil {
		return nil, errors.Wrap(err, "failed to convert kube flags to rest config")
	}
	// Configure the client to use a rate limiter
	restConfig.QPS = 2
	restConfig.Burst = 5
```

**Comment by xavpaice** [comment]
Yeah, not confident at this stage that this is a good idea, given the throttling.  It seems to be faster running the requests sequentially.

**Comment by xavpaice** [comment]
I've rebased this against the main branch, with the change that increases limits for client throttling.  The newer results show a dramatic difference between 0.45 and main, then approx. 2s difference from main to this branch.  Not a massive difference, but I think the simplification of the code alone has some value.

```
xavpaice@xav-1:~/0.45$ time ./support-bundle ../support.yaml

 Collecting support bundle ⠴ cluster-resources2022/12/23 04:22:41 An analyzer failed to run: failed to read collected ceph status: open /tmp/supportbundle4063635563/support-bundle-2022-12-23T04_22_20/ceph/status.json: no such file or directory
 Collecting support bundle ⠦ cluster-resources
support-bundle-2022-12-23T04_22_20.tar.gz

real    0m20.940s
user    0m22.085s
sys     0m1.256s


xavpaice@xav-1:~$ ./support-bundle version
Replicated Troubleshoot v0.50.0-6-ge2053a0
xavpaice@xav-1:~$ time ./support-bundle support.yaml

 Collecting support bundle ⠹ cluster-info2022/12/23 04:23:32 An analyzer failed to run: failed to read collected ceph status: open /tmp/supportbundle1290324149/support-bundle-2022-12-23T04_23_24/ceph/status.json: no such file or directory
 Collecting support bundle ⠸ cluster-infosupport-bundle-2022-12-23T04_23_24.tar.gz

real    0m8.653s
user    0m22.848s
sys     0m1.524s


xavpaice@xav-1:~$ ./support-bundle-concurrent version
Replicated Troubleshoot v0.50.0-10-gf63f51a-dirty
xavpaice@xav-1:~$ time ./support-bundle-concurrent support.yaml

 Collecting support bundle ⠙ cluster-info2022/12/23 04:24:49 An analyzer failed to run: failed to read collected ceph status: open /tmp/supportbundle1382103550/support-bundle-2022-12-23T04_24_43/ceph/status.json: no such file or directory
 Collecting support bundle ⠹ cluster-infosupport-bundle-2022-12-23T04_24_43.tar.gz

real    0m6.431s
user    0m17.970s
sys     0m1.022s
```

File: pkg/collect/cluster_resources.go (hunk 1 above)
Comment (xavpaice) [reply pkg/collect/cluster_resources.go]: moved to a constant

File: pkg/collect/cluster_resources.go [hunk 3]
```diff
@@ -156,185 +167,141 @@ func (c *CollectClusterResources) Collect(progressChan chan<- interface{}) (Coll
...
-	}
-	output.SaveResult(c.BundlePath, "cluster-resources/network-policy-errors.json", marshalErrors(networkPolicyErrors))
+	})
 
-	// resource quotas
-	resourceQuota, resourceQuotaErrors := resourceQuota(ctx, client, namespaceNames)
-	for k, v := range resourceQuota {
-		output.SaveResult(c.BundlePath, path.Join("cluster-resources/resource-quotas", k), bytes.NewBuffer(v))
+	// these collect functions all have the exact same args and structure to run them
+	apiQueries := []apiQuery{
```
Comment (banjoh) [inline pkg/collect/cluster_resources.go:203]: Nice!

File: pkg/collect/cluster_resources.go [hunk 4]
```diff
@@ -156,185 +167,141 @@ func (c *CollectClusterResources) Collect(progressChan chan<- interface{}) (Coll
...
-			podLogs, err := savePodLogs(ctx, c.BundlePath, client, &pod, "", container.Name, limits, false, false)
-			if err != nil {
-				errPath := filepath.Join("cluster-resources", "pods", "logs", pod.Namespace, pod.Name, fmt.Sprintf("%s-logs-errors.log", container.Name))
-				output.SaveResult(c.BundlePath, errPath, bytes.NewBuffer([]byte(err.Error())))
+	clusterResourceCollectors = append(clusterResourceCollectors, func() {
+		pods, podErrors, unhealthyPods := pods(ctx, client, namespaceNames)
+		for k, v := range pods {
+			output.SaveResult(c.BundlePath, path.Join("cluster-resources/pods", k), bytes.NewBuffer(v))
+		}
+		output.SaveResult(c.BundlePath, "cluster-resources/pods-errors.json", marshalErrors(podErrors))
```
Comment (banjoh) [inline pkg/collect/cluster_resources.go:178]: `CollectResult` is not thread safe

**Comment by xavpaice** [comment]
I'm going to park this for now, partly because the goroutine is not thread safe, but also that we have higher priorities.

**Pull request closed without merging**